### PR TITLE
vlinsert: journald ingestion support

### PR DESCRIPTION
**Pull request by AndrewChubatiuk**
fixes https://github.com/VictoriaMetrics/VictoriaMetrics/issues/4618

~requires https://github.com/valyala/gozstd/pull/64 to fix ZSTD compression for future systemd support~ (merged)

Added support of HTTP headers and ZSTD compression as well, though it's not [yet](https://github.com/systemd/systemd/pull/34595) supported in systemd-journal-upload. 

### Checklist

The following checks are **mandatory**:

- [ ] My change adheres [VictoriaMetrics contributing guidelines](https://docs.victoriametrics.com/contributing/).

### CODE REVIEW

File: app/vlinsert/journald/journald.go [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+		return
+	}
+	if err := vlstorage.CanWriteData(); err != nil {
+		httpserver.Errorf(w, r, "%s", err)
+		return
+	}
+
+	tenantID, err := logstorage.ParseTenantID(*journaldTenantID)
+	if err != nil {
+		httpserver.Errorf(w, r, "cannot parse -journald.tenantID=%q for journald: %s", *journaldTenantID, err)
```
Comment (f41gh7) [inline app/vlinsert/journald/journald.go:92]: looks like return is missing

File: app/vlinsert/journald/journald.go [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+	if err != nil {
+		httpserver.Errorf(w, r, "cannot read request body: %s", err)
+		return
+	}
+
+	if err != nil {
+		httpserver.Errorf(w, r, "cannot parse common params from request: %s", err)
+		return
+	}
+	if err := vlstorage.CanWriteData(); err != nil {
```
Comment (f41gh7) [inline app/vlinsert/journald/journald.go:85]: It'd be great to move this check before reading request body with io.ReadAll

File: app/vlinsert/journald/journald.go [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+var (
+	rowsIngestedJournaldTotal = metrics.NewCounter(`vl_rows_ingested_total{type="journald", format="journald"}`)
+
+	requestsJournaldTotal = metrics.NewCounter(`vl_http_requests_total{path="/insert/journald/upload",format="journald"}`)
+	errorsTotal           = metrics.NewCounter(`vl_http_errors_total{path="/insert/journald/upload",format="journald"}`)
+
+	requestJournaldDuration = metrics.NewHistogram(`vl_http_request_duration_seconds{path="/insert/journald/upload",format="journald"}`)
+)
+
+func parseJournaldRequest(data []byte, lmp insertutils.LogMessageProcessor, cp *insertutils.CommonParams) (int, error) {
```
Comment (f41gh7) [inline app/vlinsert/journald/journald.go:121]: Could you please add test for `parseJournaldRequest` function?

**Review by f41gh7 — CHANGES_REQUESTED** [review]
Overall looks good!

 Please see comments

**Comment by AndrewChubatiuk** [comment]
@f41gh7 updated PR, please take a look

File: app/vlinsert/journald/journald.go [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	dataStr := bytesutil.ToUnsafeString(data)
+	var fields []logstorage.Field
+	var ts int64
+	var err error
+	var rowsIngested, lastIdx int
+	var name, value string
+	for i, char := range dataStr {
```
Comment (f41gh7) [inline app/vlinsert/journald/journald.go:144]: This code has performance issue - it iterates over string char, which is very slow.

 It's better to use index.Byte() by new line separator \n

File: app/vlinsert/journald/journald.go [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+var (
+	rowsIngestedJournaldTotal = metrics.NewCounter(`vl_rows_ingested_total{type="journald", format="journald"}`)
+
+	requestsJournaldTotal = metrics.NewCounter(`vl_http_requests_total{path="/insert/journald/upload",format="journald"}`)
+	errorsTotal           = metrics.NewCounter(`vl_http_errors_total{path="/insert/journald/upload",format="journald"}`)
+
+	requestJournaldDuration = metrics.NewHistogram(`vl_http_request_duration_seconds{path="/insert/journald/upload",format="journald"}`)
+)
+
+func parseJournaldRequest(data []byte, lmp insertutils.LogMessageProcessor, cp *insertutils.CommonParams) (int, error) {
```
Comment (f41gh7) [inline app/vlinsert/journald/journald.go:137]: Please add comment with link to the format documentation https://systemd.io/JOURNAL_EXPORT_FORMATS/#journal-export-format

 Also, maybe it's better to use JSON as journald encoding? It must simplify parsing.

File: app/vlinsert/journald/journald_test.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+	f("__REALTIME_TIMESTAMP=91723819283\nMESSAGE=Test message",
+		[]int64{91723819283000},
+		"{\"_msg\":\"Test message\"}",
+	)
+
+	// Multiple events
+	f("__REALTIME_TIMESTAMP=91723819283\nMESSAGE=Test message\n\n__REALTIME_TIMESTAMP=91723819284\nMESSAGE=Test message2",
+		[]int64{91723819283000, 91723819284000},
+		"{\"_msg\":\"Test message\"}\n{\"_msg\":\"Test message2\"}",
+	)
```
Comment (f41gh7) [inline app/vlinsert/journald/journald_test.go:36]: Please add the following test case: 
```
_CURSOR=s=bcce4fb8ffcb40e9a6e05eee8b7831bf;i=5ef603;b=ec25d6795f0645619ddac9afdef453ee;m=545242e7049;t=50f1202
__REALTIME_TIMESTAMP=1423944916375353
__MONOTONIC_TIMESTAMP=5794517905481
_BOOT_ID=ec25d6795f0645619ddac9afdef453ee
_TRANSPORT=journal
_UID=1001
_GID=1001
_CAP_EFFECTIVE=0
_SYSTEMD_OWNER_UID=1001
_SYSTEMD_SLICE=user-1001.slice
_MACHINE_ID=5833158886a8445e801d437313d25eff
_HOSTNAME=bupkis
_AUDIT_LOGINUID=1001
_SELINUX_CONTEXT=unconfined_u:unconfined_r:unconfined_t:s0-s0:c0.c1023
CODE_LINE=1
CODE_FUNC=<module>
SYSLOG_IDENTIFIER=python3
_COMM=python3
_EXE=/usr/bin/python3.4
_AUDIT_SESSION=35898
_SYSTEMD_CGROUP=/user.slice/user-1001.slice/session-35898.scope
_SYSTEMD_SESSION=35898
_SYSTEMD_UNIT=session-35898.scope
MESSAGE
^G^@^@^@^@^@^@^@foo
bar
CODE_FILE=<string>
_PID=16853
_CMDLINE=python3 -c from systemd import journal; journal.send("foo\nbar")
_SOURCE_REALTIME_TIMESTAMP=1423944916372858
```

Comment (AndrewChubatiuk) [reply app/vlinsert/journald/journald_test.go:36]: event's message is not valid, I've tested it locally

File: app/vlinsert/journald/journald.go (hunk 5 above)
Comment (AndrewChubatiuk) [reply app/vlinsert/journald/journald.go:137]: json encoding is only supported by journal gateway, which requires to switch victorialogs to pull model

File: app/vlinsert/journald/journald_test.go (hunk 6 above)
Comment (f41gh7) [reply app/vlinsert/journald/journald_test.go:36]: It's valid, according to the doc it should be encoding the binary data:
```
Other journal fields are serialized in a special binary safe way: field name, followed by newline, followed by a binary 64-bit little endian size value, followed by the binary field data, followed by a newline as separator to the next field.
```

 See this doc example:

```
python3 -c 'from systemd import journal; journal.send("foo\nbar")'
journalctl -n1 -o export
```

Comment (AndrewChubatiuk) [reply app/vlinsert/journald/journald_test.go:36]: I've executed it locally, but had no
```
MESSAGE
^G^@^@^@^@^@^@^@foo
bar
```

Comment (AndrewChubatiuk) [reply app/vlinsert/journald/journald_test.go:36]: I received this instead
```
__CURSOR=s=e0afe8412a6a49d2bfcf66aa7927b588;i=1e43;b=f778b6e2f7584a77b991a2366612a7b5;m=2ebc1b234c;t=6252591539469;x=5aaf020c6557a0d4
__REALTIME_TIMESTAMP=1729693142455401
__MONOTONIC_TIMESTAMP=200724390732
__SEQNUM=7747
__SEQNUM_ID=e0afe8412a6a49d2bfcf66aa7927b588
_BOOT_ID=f778b6e2f7584a77b991a2366612a7b5
_UID=0
_GID=0
_MACHINE_ID=a4a970370c30a925df02a13c67167847
_HOSTNAME=ecd5e4555787
_RUNTIME_SCOPE=system
_TRANSPORT=journal
_CAP_EFFECTIVE=1ffffffffff
_SYSTEMD_CGROUP=/init.scope
_SYSTEMD_UNIT=init.scope
_SYSTEMD_SLICE=-.slice
MESSAGE
foo
bar
CODE_FILE=<stdin>
CODE_LINE=1
CODE_FUNC=<module>
SYSLOG_IDENTIFIER=python3
_COMM=python3
_EXE=/usr/bin/python3.12
_CMDLINE=python3
_PID=2676
_SOURCE_REALTIME_TIMESTAMP=1729693142455200
```

Comment (f41gh7) [reply app/vlinsert/journald/journald_test.go:36]: parser implementation with C https://github.com/systemd/systemd/blob/main/src/shared/journal-importer.c#L369

Comment (AndrewChubatiuk) [reply app/vlinsert/journald/journald_test.go:36]: updated

**Review by f41gh7 — APPROVED** [review]
LGTM

**Comment by f41gh7** [comment]
Thanks for contribution!

**Pull request merged**